From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peter Pan <peter@never.lan>
Date: Thu, 28 May 2026 12:06:43 -0400
Subject: [PATCH] Fix Paper Version String


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index daccd8fba0953d6b62034d0d3c4c70a774cb04b8..4d87acfeb587d9ef3039017ddb6cedb6efaaf048 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -650,7 +650,7 @@ public final class CraftServer implements Server {
 
     @Override
     public String getVersion() {
-        return this.serverVersion + " (MC: " + this.console.getServerVersion() + ")";
+        return this.bukkitVersion;
     }
 
     @Override
