From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peter Pan <peter@never.lan>
Date: Tue, 9 Jun 2026 22:05:54 -0400
Subject: [PATCH] Useful Raids


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftRaid.java b/src/main/java/org/bukkit/craftbukkit/CraftRaid.java
index c4d4e2b35ea9dbf3380bec8f899db722f915377f..599269e20ad0b8cab007e867475ffa6e659f4d43 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftRaid.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftRaid.java
@@ -114,6 +114,18 @@ public final class CraftRaid implements Raid {
         return this.handle.idOrNegativeOne;
     }
 
+    // Aincrad start - more Raid API
+    @Override
+    public void forceStop(boolean victory) {
+        this.handle.forceStop((net.minecraft.server.level.ServerLevel) this.level, victory);
+    }
+
+    @Override
+    public boolean isPluginControlled() {
+        return this.handle.pluginControlled;
+    }
+    // Aincrad end - more Raid API
+
     @Override
     public BossBar getBossBar() {
         return new CraftBossBar(this.handle.raidEvent);
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index ea75f1bf46624ddf2a4e95cd031729f958fe9e77..6bcd0bb09ab678810f50fe6ecb898de5f571f499 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -1920,6 +1920,16 @@ public class CraftWorld extends CraftRegionAccessor implements World {
     }
     // Paper end - more Raid API
 
+    // Aincrad start - more Raid API
+    @Override
+    public @Nullable Raid createRaid(@org.jetbrains.annotations.NotNull org.bukkit.entity.Player player, @org.jetbrains.annotations.NotNull org.bukkit.Location location, int badOmenLevel, boolean pluginControlled) {
+        net.minecraft.server.level.ServerPlayer serverPlayer = ((org.bukkit.craftbukkit.entity.CraftPlayer) player).getHandle();
+        net.minecraft.core.BlockPos pos = org.bukkit.craftbukkit.util.CraftLocation.toBlockPosition(location);
+        net.minecraft.world.entity.raid.Raid raid = this.world.getRaids().createRaid(serverPlayer, pos, badOmenLevel, pluginControlled);
+        return raid != null ? new CraftRaid(raid, this.world) : null;
+    }
+    // Aincrad end - more Raid API
+
     @Override
     public List<Raid> getRaids() {
         Raids persistentRaid = this.world.getRaids();
